metrics/operator_repo_stats.py

```diff
@@ -0,0 +1,93 @@
+"""Module for parsing operator repositories into useful metrics."""
+
+import time
+from pathlib import Path
+import subprocess
+from threading import Thread, Event
+
+import yaml
+
+from metrics.prometheus_metrics import (
+    MIGRATION_GAUGE,
+    BUNDLES_IN_REPO_GAUGE,
+    OPERATORS_IN_REPO_GAUGE,
+    MIGRATION_COUNT_GAUGE,
+)
+
+
+CLONE_DIR = Path("/tmp")  # nosec
+OPERATORS_DIR_NAME = "operators"
+CI_FILE_NAME = "ci.yaml"
+SYNC_DELAY = 86400  # 24 h in seconds
+
+
+def ensure_repo(repo: Path, git_url: str) -> None:
+    """Ensure a local repository exists and is up to date."""
+    if repo.exists():
+        pop = subprocess.Popen(
+            ["cd", repo.absolute(), ";", "git", "pull"], shell=True
+        )  # nosec
+    else:
+        # pylint: disable=consider-using-with
+        pop = subprocess.Popen(["git", "clone", git_url, repo])
+        # pylint: enable=consider-using-with
+    pop.wait()
+
+
+def load_configured_repos(config_path: Path | str = "repos.yml") -> dict[str, str]:
+    """Load config file with repository names and remotes."""
+    with open(config_path, "r", encoding="utf-8") as i_file:
+        data: dict[str, str] = yaml.safe_load(i_file)
+        return data
+
+
+def parse_stats(repo: Path) -> None:
+    """Parses stats for a single operator repository."""
+    operators_dir = repo / OPERATORS_DIR_NAME
+    if operators_dir.exists() and operators_dir.is_dir():
+        num_of_operators = 0
+        migrated_operators = 0
+        for operator in operators_dir.iterdir():
+            if not operator.is_dir():
+                # The directory can also contain files like README
+                continue
+            ci_file = operator / CI_FILE_NAME
+            if not ci_file.exists():
+                continue
+            # This is an operator
+            num_of_operators += 1
+            # Update number of bundles
+            BUNDLES_IN_REPO_GAUGE.labels(
```

Comment:
Not all subdirectories inside the operator directories are bundles. There is for example a directory for catalog-templates. If you don't want to be bothered by the implementing your own detection logic I would recoment using https://github.com/mporrato/operator-repo library that gives you easy access interface for an operator repository and handles all this logic automatically.

Reply:
With this library, you can initiate `Repo` object (https://github.com/mporrato/operator-repo/blob/main/src/operator_repo/core.py#L818) and get access to all operators and bundles using Python classes without need or reading directories and files.

Reply:
Great suggestion, will do that. :+1: